qcodes/instrument_drivers/rigol/DS1074Z.py

```diff
@@ -0,0 +1,235 @@
+from typing import Union
+import numpy as np
+
+
+from qcodes import VisaInstrument
+from qcodes import InstrumentChannel
+from qcodes import Instrument
+from qcodes import ChannelList
+from qcodes import ParameterWithSetpoints, Parameter
+from qcodes.utils.validators import Numbers, Enum
+from qcodes.utils.validators import Arrays
+
+
+class TraceNotReady(Exception):
+    pass
+
+
+class ScopeTrace(ParameterWithSetpoints):
+
+    def __init__(self, name, channel, **kwargs):
+        super().__init__(name, **kwargs)
+        self._trace_ready = False
+        self._channel = channel
+
+    def prepare_curvedata(self):
+        """
+        Prepare oscilloscope to return the trace
+        """
+        npts = self.root_instrument.waveform_npoints()
+        self.shape = (npts,)
+        self._trace_ready = True
```

Comment:
I may be missing something important, but I don't think this dance of preparing the trace is needed when using `ParameterWithSetpoints`. I think it's an artifact of how `ArrayParameter` works.

Reply:
@WilliamHPNielsen Yes that was the major motivation for `ParameterWithSetpoints` to get rid of this

Reply:
@William You were indeed correct, all the dance was not needed. I have simplified the code i.e. remove the unnecessary ScopeTrace(ParameterWithSetpoints) class and implement the get_trace directly in the RigolDS1074ZChannel(InstrumentChannel) class. 

commit name: Remove ScopeTrace and place methods in RigolDS1074ZChannel
commit hash: 35fa366a2b65a586f2772776a078160342dcebb1